Fluid dynamics/stokes for prism hex quad

mohittyagi-k opened the pull request:
Hi All,

**Description**

This pull request is for the enabling symbolic stokes element for prism and hex; and stokes wall condition on a quad. 
I need some clarifications in the changes, and I would need your help on how to go about it.

### Key points:

- to enable stokes for prism, hex
- derived from FluidElement
- the template is similar to that for two-fluid-Navier-stokes, the only difference is that the advection, enrichment, and darcy terms are removed
- added symbolic stokes data
- added StokesWallCondition3D4N for quad
- added minimum and average element size computation for prism

### Tasks:

- [x]  Changes in element_size_calculator - PR https://github.com/KratosMultiphysics/Kratos/pull/7224
- [x] Tests update - use of new macros, and remove deprecated interface call for `Initialize`, add check for LHS
- [x] Implementation of `Calculate` method for FLUID_STRESS in `FluidElement` -PR https://github.com/KratosMultiphysics/Kratos/pull/7181
- [x] A common generator for Stokes and NavierStokes - PR https://github.com/KratosMultiphysics/Kratos/pull/7240, and will be merged before this PR -- changes merged in this PR.
- [x] Add simple cpp tests for the quad, prism, hex

---------------------------------------------------------------------------------------------------------------------------
@ddiezrod @rubenzorrilla  @pooyan-dadvand  @KratosMultiphysics/fluid-dynamics @KratosMultiphysics/altair

Commit:
a general stokes element

Commit:
lyx modifications

Commit:
prism3d6 - minimum element size computation

Commit:
prism3d6 - average element size computation

Commit:
Using SymbolicStokes in the name for the Stokes element

Commit:
File not needed

Commit:
Description correction

Commit:
Description correction

Commit:
Removing unsued variable

Commit:
Remove enrichment shape function

Commit:
Remove not required file

Commit:
Removing unused variables from template

Commit:
Update stokes element

Commit:
Stokes wall condition for quad

Commit:
comment correction

Commit:
Bugfix: (Also) Update material response at IP

- template and element update

Commit:
fix in the minimum size calculation for prism

Commit:
modifying template for correct substitution

Commit:
appropriate substitution based on geometry type

Commit:
updated stokes element

Commit:
normal computation correction for quad

Commit:
register stokes wall condition for quad

Commit:
adding symbolic stokes data

Commit:
symbolic stokes data based intantiation for stokes

Commit:
Refactor name from stokes to symbolic stokes

Commit:
some more cleanup

Commit:
tau1 adjustment for stationary stokes

Commit:
updated stokes elemnt

Commit:
test for symbolic stokes tet

Commit:
Correction for tet

Commit:
test for prism and hex

Commit:
adding point of contact for these files

Commit:
throwing error for non implemented function

- not required for symbolic stokes

mohittyagi-k commented on applications/FluidDynamicsApplication/custom_conditions/stokes_wall_condition.cpp:
```diff
@@ -144,7 +209,7 @@ void StokesWallCondition<TDim,TNumNodes>::ApplyNeumannCondition(MatrixType &rLoc
     Normal /= A;
 
-    // CAUTION: "Jacobian" is 2.0*A for triangles but 0.5*A for lines 
+    // CAUTION: "Jacobian" is 2.0*A for triangles but 0.5*A for lines
```
I have a doubt here. Is Jacobian 0.25 * A for quad?

mohittyagi-k replied:
Now using the method from geometry to compute the Jacobian determinant, so this doubt is no longer relevant.

mohittyagi-k commented on applications/FluidDynamicsApplication/custom_conditions/stokes_wall_condition.cpp:
```diff
@@ -129,6 +177,23 @@ void StokesWallCondition<3,3>::CalculateNormal(array_1d<double,3>& An )
 }
 
+template <>
+void StokesWallCondition<3,4>::CalculateNormal(array_1d<double,3>& An )
```
I have implemented this method to compute Normal (times Area) for quad.

philbucher replied:
this should already be implemented in the Geometry itself, can you check please?

mohittyagi-k replied:
yes, Normal function is in Geometry. Removed this function.

ddiezrod commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp:
```diff
@@ -0,0 +1,4943 @@
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main author:     Daniel Diez
```
Not me : P

mohittyagi-k replied:
Is it fine now? If not, please suggest changes. 😄

mohittyagi-k replied:
NA

ddiezrod commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp:
```diff
@@ -0,0 +1,4943 @@
+    {
+        dt_inv = 1.0/dt;
+    }
+    if (fabs(bdf0) < 1e-9)
```
better use `std::abs ` in c++

mohittyagi-k replied:
done

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_elements/fluid_element.cpp:
```diff
@@ -807,6 +808,12 @@ void StrainRateSpecialization<TElementData,3>::Calculate(
+template class FluidElement< SymbolicStokesData<2,3> >;
+template class FluidElement< SymbolicStokesData<2,4> >;
+template class FluidElement< SymbolicStokesData<3,4> >;
+template class FluidElement< SymbolicStokesData<3,6> >;
```
Why `<3,6>`? Does it makes sense to have wedge elements in the context of CFD?

mrhashemi replied:
I agree with @rubenzorrilla 
Wedge elements are (_usually_) utilized when directionality matters, e.g. modeling shells, or some Darcy flow cases.

RiccardoRossi replied:
it is also positive for Boundary Layers, so i think the prism is an important element

rubenzorrilla replied:
OK. Forget about my comment @mohittyagi-k and my apologies for the noise. Thanks for the explanation @RiccardoRossi .

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp:
```diff
@@ -0,0 +1,4943 @@
+    } else{
+        KRATOS_ERROR << "SymbolicStokes is supposed to manage time integration." << std::endl;
+    }
+}
```
The implementation in the base class can be used.

mohittyagi-k replied:
agree, done.

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp:
```diff
@@ -0,0 +1,4943 @@
+{
+    MatrixType tmp;
+    CalculateLocalSystem(tmp, rRightHandSideVector, rCurrentProcessInfo);
+}
```
This is an overhead. The `CalculateRightHandSide` can be easily implemented with the automatic differentiation.

mohittyagi-k replied:
Now calling the base CalculateRightHandSide. This should avoid the problem of overhead.

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp:
```diff
@@ -0,0 +1,4943 @@
+        << "Error in base class Check for Element " << this->Info() << std::endl
+        << "Error code is " << out << std::endl;
+
+    KRATOS_CHECK_VARIABLE_KEY( DIVERGENCE );
```
After the new registering of variables mechanism this is no longer required. @pooyan-dadvand  can you confirm?

Hence, the implementation in the base class can be used.

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp:
```diff
@@ -0,0 +1,4943 @@
+
+        for (unsigned int i = 0; i < StrainSize; i++){
+            rOutput[i] = ( 1.0 / sumOfGaussWeights ) * rOutput[i];
+        }
```
This should be implemented in the base class (not done yet).

mohittyagi-k replied:
Now, this is implemented in the base class (in https://github.com/KratosMultiphysics/Kratos/pull/7181)

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp:
```diff
@@ -0,0 +1,4943 @@
+{
+    rData.UpdateGeometryValues(IntegrationPointIndex, Weight, rN, rDN_DX);
+    this->CalculateMaterialResponse(rData);
+}
```
Copy and paste from base class. Please remove.

mohittyagi-k replied:
done.

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp:
```diff
@@ -0,0 +1,4943 @@
+    if (fabs(bdf0) < 1e-9)
+    {
+        dt_inv = 0.0;
+    }
```
You are checking twice the same statement. Please use an if & else statement.

rubenzorrilla replied:
Besides, I don't think this needs to be checked for each element.

mohittyagi-k replied:
> You are checking twice the same statement. Please use an if & else statement.

They are meant to be two different checks. Sorry for the confusion, have added a comment now.


> Besides, I don't think this needs to be checked for each element.

Agree, but how to do it.

mohittyagi-k replied:
NA

philbucher commented on kratos/utilities/element_size_calculator.cpp:
```diff
@@ -225,6 +293,26 @@ double ElementSizeCalculator<3,4>::AverageElementSize(const Geometry<Node<3> >&
+    double z30 = rGeometry[3].Z() - rGeometry[0].Z();
+
+    double detJ = 0.5 * (x10 * y20 * z30 - x10 * y30 * z20 + y10 * z20 * x30 - y10 * x20 * z30 + z10 * x20 * y30 - z10 * y20 * x30);
+    return pow(detJ, 1. / 3.);
```
```suggestion
    return std::pow(detJ, 1. / 3.);
```

mohittyagi-k replied:
done.

philbucher commented on kratos/utilities/element_size_calculator.cpp:
```diff
@@ -138,6 +138,74 @@ double ElementSizeCalculator<3,4>::MinimumElementSize(const Geometry<Node<3> >&
+    // n_v21 /= || n_v21||
+
+    // Projection from Node 0 on edge 12:
+    // Hsq = v10 dot n_v21
```
```suggestion
```

please remove commented code

mohittyagi-k replied:
done

philbucher commented on applications/FluidDynamicsApplication/symbolic_generation/stokes/stokes_element.lyx:
```diff
@@ -0,0 +1,1005 @@
+#LyX 2.2 created this file. For more info see http://www.lyx.org/
```
what is this file? Does it need to be in the repo?

mohittyagi-k replied:
I guess this file is good for documentation purposes. You may open it using LyX software. I saw these similar files in a couple of other elements, so thought of keeping it.

philbucher replied:
if it is useful in the future / for documentation then I would add it to the [Documentation Repo](https://github.com/KratosMultiphysics/Documentation)
I just checked there is nothing from FluidDynamics there yet but I think it would be a good start
Do you agree @rubenzorrilla ?

rubenzorrilla replied:
I think this is a good idea. I don't like pretty much mixing the code with the documentation. In the README.md of the application we can add hlinks to the documentation repo.

mohittyagi-k replied:
.lyx file is no more in this PR. So resolving the conversation.

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp:
```diff
@@ -0,0 +1,4943 @@
+}
+
+template <>
+void SymbolicStokes<SymbolicStokesData<2,3>>::ComputeGa
```
```suggestion
void SymbolicStokes<TElementData>::GetValueOnIntegrationPoints(
    const Variable<double> &rVariable,
    std::vector<double> &rValues,
    const ProcessInfo &rCurrentProcessInfo)
```
Please follow the style guide.

mohittyagi-k replied:
Done.

philbucher commented on applications/FluidDynamicsApplication/tests/cpp_tests/test_symbolic_stokes_element.cpp:
```diff
@@ -0,0 +1,328 @@
+            Vector RHS = ZeroVector(24);
+            Matrix LHS = ZeroMatrix(24, 24);
+            p_element->Initialize();
+            p_element->CalculateLocalSystem(LHS, RHS, process_info);
```
why do you call `CalculateLocalSystem` and only check the RHS?

mohittyagi-k replied:
agree, on it.

mohittyagi-k replied:
done

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp:
```diff
@@ -0,0 +1,4943 @@
+}
+
+template <>
+void SymbolicStokes<SymbolicStokesData<2,3>>::ComputeGa
```
If the variable is not in the element the base class calculate must be called.

As a comment aside I think we should at some point implement the `DIVERGENCE` in the base class.

philbucher commented on applications/FluidDynamicsApplication/tests/cpp_tests/test_symbolic_stokes_element.cpp:
```diff
@@ -0,0 +1,328 @@
+            KRATOS_CHECK_NEAR(RHS(20), -100, 1e-07);
+            KRATOS_CHECK_NEAR(RHS(21), -108.3333333, 1e-07);
+            KRATOS_CHECK_NEAR(RHS(22), -147.2222222, 1e-07);
+            KRATOS_CHECK_NEAR(RHS(23), -0.05520833333, 1e-07);
```
please use the check macros for [vectors and matrices](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/checks.h#L53-L151)
Macro unrolling is expensive and increases compilation time

mohittyagi-k replied:
done

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.h:
```diff
@@ -0,0 +1,405 @@
+     */
+    void GetValueOnIntegrationPoints(   const Variable<double> &rVariable,
+                                        std::vector<double> &rValues,
+                                        const ProcessInfo &rCurrentProcessInfo ) override;
```
```suggestion
    void GetValueOnIntegrationPoints(
        const Variable<double> &rVariable,
        std::vector<double> &rValues,
        const ProcessInfo &rCurrentProcessInfo) override;
```

rubenzorrilla replied:
Please follow the style guide.

mohittyagi-k replied:
done

philbucher commented on applications/FluidDynamicsApplication/tests/cpp_tests/test_symbolic_stokes_element.cpp:
```diff
@@ -0,0 +1,328 @@
+            Vector RHS = ZeroVector(32);
+            Matrix LHS = ZeroMatrix(32, 32);
+            p_element->Initialize();
+            p_element->CalculateLocalSystem(LHS, RHS, process_info);
```
same

philbucher commented on applications/FluidDynamicsApplication/tests/cpp_tests/test_symbolic_stokes_element.cpp:
```diff
@@ -0,0 +1,328 @@
+            p_element->Initialize();
+            p_element->CalculateLocalSystem(LHS, RHS, process_info);
+            // Check the RHS values
+            KRATOS_CHECK_NEAR(RHS(0), -29.16666667, 1e-07);
```
same

mohittyagi-k replied:
done

philbucher commented on applications/FluidDynamicsApplication/tests/cpp_tests/test_symbolic_stokes_element.cpp:
```diff
@@ -0,0 +1,328 @@
+            // Compute RHS and LHS
+            Vector RHS = ZeroVector(32);
+            Matrix LHS = ZeroMatrix(32, 32);
+            p_element->Initialize();
```
```suggestion
            p_element->Initialize(process_info);
```

you were using the deprecated interface

mohittyagi-k replied:
done.

philbucher commented on applications/FluidDynamicsApplication/tests/cpp_tests/test_symbolic_stokes_element.cpp:
```diff
@@ -0,0 +1,328 @@
+            // Compute RHS and LHS
+            Vector RHS = ZeroVector(24);
+            Matrix LHS = ZeroMatrix(24, 24);
+            p_element->Initialize();
```
```suggestion
            p_element->Initialize(process_info);

```

philbucher commented on applications/FluidDynamicsApplication/tests/cpp_tests/test_symbolic_stokes_element.cpp:
```diff
@@ -0,0 +1,328 @@
+
+            // Set the nodal DENSITY and DYNAMICS_VISCOSITY
+            Element::Pointer p_element = model_part.pGetElement(1);
+            for (NodeIteratorType it_node = model_part.NodesBegin(); it_node != model_part.NodesEnd(); ++it_node)
```
```suggestion
            for (auto& r_node: model_part.Nodes())

```

philbucher commented on applications/FluidDynamicsApplication/tests/cpp_tests/test_symbolic_stokes_element.cpp:
```diff
@@ -0,0 +1,328 @@
+            // Compute RHS and LHS
+            Vector RHS = ZeroVector(16);
+            Matrix LHS = ZeroMatrix(16, 16);
+            p_element->Initialize();
```
```suggestion
            p_element->Initialize(process_info);
```

philbucher commented on applications/FluidDynamicsApplication/tests/cpp_tests/test_symbolic_stokes_element.cpp:
```diff
@@ -0,0 +1,328 @@
+            Vector RHS = ZeroVector(16);
+            Matrix LHS = ZeroMatrix(16, 16);
+            p_element->Initialize();
+            p_element->CalculateLocalSystem(LHS, RHS, process_info);
```
same

philbucher commented on applications/FluidDynamicsApplication/tests/cpp_tests/test_symbolic_stokes_element.cpp:
```diff
@@ -0,0 +1,328 @@
+            p_element->Initialize();
+            p_element->CalculateLocalSystem(LHS, RHS, process_info);
+            // Check the RHS values
+            KRATOS_CHECK_NEAR(RHS(0), 9.722222285, 1e-07);
```
same

philbucher commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.h:
```diff
@@ -0,0 +1,405 @@
+    void CalculateLocalSystem(
+        MatrixType &rLeftHandSideMatrix,
+        VectorType &rRightHandSideVector,
+        ProcessInfo &rCurrentProcessInfo) override;
```
```suggestion
        const ProcessInfo &rCurrentProcessInfo) override;
```

philbucher commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.h:
```diff
@@ -0,0 +1,405 @@
+     */
+    void CalculateRightHandSide(
+        VectorType &rRightHandSideVector,
+        ProcessInfo &rCurrentProcessInfo) override;
```
```suggestion
        const ProcessInfo &rCurrentProcessInfo) override;
```

mohittyagi-k replied:
Now calling the base class function to compute it.

philbucher commented on applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.h:
```diff
@@ -0,0 +1,405 @@
+     * current element check implementations
+     * @param rCurrentProcessInfo reference to the current process info
+     */
+    int Check(const ProcessInfo &rCurrentProcessInfo) override;
```
```suggestion
    int Check(const ProcessInfo &rCurrentProcessInfo) const override;
```

mohittyagi-k replied:
Removed the Check. Now using the base class Check, which has been changed to const.

philbucher commented on applications/FluidDynamicsApplication/symbolic_generation/stokes/symbolic_stokes_template.cpp:
```diff
@@ -0,0 +1,764 @@
+        << "Error in base class Check for Element " << this->Info() << std::endl
+        << "Error code is " << out << std::endl;
+
+    KRATOS_CHECK_VARIABLE_KEY( DIVERGENCE );
```
```suggestion
```

the error is thrown somewhere else, you can just remove this

mohittyagi-k replied:
done.

philbucher commented on applications/FluidDynamicsApplication/symbolic_generation/stokes/symbolic_stokes_template.cpp:
```diff
@@ -0,0 +1,764 @@
+    {
+        dt_inv = 1.0/dt;
+    }
+    if (fabs(bdf0) < 1e-9)
```
```suggestion
    if (std::abs(bdf0) < 1e-9)
```

mohittyagi-k replied:
done

philbucher commented on applications/FluidDynamicsApplication/symbolic_generation/stokes/symbolic_stokes_template.cpp:
```diff
@@ -0,0 +1,764 @@
+    {
+        dt_inv = 1.0/dt;
+    }
+    if (fabs(bdf0) < 1e-9)
```
```suggestion
    if (std::abs(bdf0) < 1e-9)
```

mohittyagi-k replied:
done.

philbucher reviewed (changes requested):
Hi @mohittyagi-k 
Overall nice work, I made a bunch of comments

for additions to the core please also add tests
in this case it should be a small C++ test

rubenzorrilla reviewed (changes requested):
I've several concerns about this PR:
- Several  methods have the same functionality as the ones in the base `FluidElement`.
- The symbolic generation file is exactly the same as the Navier-Stokes one (the two fluids was created from this one too). I am strongly against adding a complete new generator just to remove the convective term from that one.
- I think that the Stokes condition is not required and the already existent Navier-Stokes one can be used, which has much more functionalities ( mainly Neumann BC, Behr contribution, Navier-slip wall law and outlet inflow prevention). Having said this, I suggest starting to use that one rather than enhancing the Stokes one.
- Some efficiency issues.

Commit:
remove the commented code

Commit:
specify std namespace for pow

Commit:
use std::abs instead of fabs

Commit:
exact same functions definition available in base class, so removing

Commit:
style change

Commit:
calling const Check function

Commit:
commenting unused variable stab_c2

Commit:
fix build error on ubuntu

- calling functions on this

Commit:
Removing Check function from symbolic stokes.

- relying on the base class Check

Commit:
Using base class function to compute right hand side, so removing it

Commit:
adding comment

Commit:
style guide change

Commit:
adding newline at the end of the file

Commit:
using to new interface

Commit:
refactor

Commit:
use of new macro to check vector near

Commit:
adding check for lhs

Commit:
removing unused variables

Commit:
possible fix for build failure on ubuntu(gcc)

Commit:
Merge branch 'master' into FluidDynamics/stokes-for-prism-hex-quad

Commit:
remove unused variables

Commit:
Merge branch 'master' into FluidDynamics/stokes-for-prism-hex-quad

Commit:
Remove Calculate as implementation moved to base class

Commit:
possible fix build failure on linux

Commit:
using geometry to compute jacobian determinant

Commit:
refactor

Commit:
Using geometry Normal function to compute normal

philbucher commented on applications/FluidDynamicsApplication/custom_conditions/stokes_wall_condition.cpp:
```diff
@@ -99,58 +122,52 @@ void StokesWallCondition<3,3>::GetDofList(DofsVectorType& rElementalDofList,
 template <>
-void StokesWallCondition<3,3>::CalculateNormal(array_1d<double,3>& An )
+void StokesWallCondition<3,4>::GetDofList(DofsVectorType& rElementalDofList,
+        ProcessInfo& rCurrentProcessInfo)
```
```suggestion
        const ProcessInfo& rCurrentProcessInfo) const
```

mohittyagi-k replied:
done

philbucher commented on applications/FluidDynamicsApplication/custom_conditions/stokes_wall_condition.cpp:
```diff
@@ -51,6 +51,29 @@ void StokesWallCondition<3,3>::EquationIdVector(EquationIdVectorType& rResult,
+ */
+template <>
+void StokesWallCondition<3,4>::EquationIdVector(EquationIdVectorType& rResult,
+        ProcessInfo& rCurrentProcessInfo)
```
```suggestion
        const ProcessInfo& rCurrentProcessInfo) const
```

mohittyagi-k replied:
done

philbucher reviewed (commented):
minor for using the new element/condition interface

Commit:
Merge branch 'master' into FluidDynamics/stokes-for-prism-hex-quad

Commit:
Merge branch 'master' into FluidDynamics/stokes-for-prism-hex-quad

# Conflicts:
#	kratos/utilities/element_size_calculator.cpp

Commit:
style guide changes

Commit:
using new interface

Commit:
adding points of contact

Commit:
Merge branch 'master' into FluidDynamics/stokes-for-prism-hex-quad

Commit:
Merge branch 'master' into FluidDynamics/stokes-for-prism-hex-quad

# Conflicts:
#	applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp
#	applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.h
#	applications/FluidDynamicsApplication/custom_utilities/symbolic_stokes_data.h

Commit:
refactor the common generator

Commit:
removing not required files

- all this has been done in another PR 7240 by rubenzorrilla
which dealt with a common generator, templates for symbolic
stokes element, symbolic stokes element and symbolic stokes data

Commit:
removing not required test

- test for symbolic stokes added in PR 7240, now merged.

Commit:
get contributions

Commit:
for prism

Commit:
register element - quad, prism, hex

philbucher reviewed (dismissed):
:+1:

Commit:
cleanup

Commit:
further clenaup

Commit:
reverting pseudo change

Commit:
removing duplicate include

Commit:
bugfix: removing duplicate explicit instantiations

Commit:
cleanup

- removing duplicate include
- reverting pseudo change

Commit:
remove newline

Commit:
remove duplicate include

Commit:
removing duplicate line

rubenzorrilla commented on applications/FluidDynamicsApplication/symbolic_generation/navier_stokes/generate_navier_stokes_element.py:
```diff
@@ -26,7 +26,7 @@
 divide_by_rho = True                # Divide the mass conservation equation by rho
 ASGS_stabilization = True           # Consider ASGS stabilization terms
-formulation = "NavierStokes"        # Element type. Options: "NavierStokes", "Stokes"
+formulation = "Stokes"        # Element type. Options: "NavierStokes", "Stokes"
```
```suggestion
formulation = "NavierStokes"        # Element type. Options: "NavierStokes", "Stokes"
```
(just to keep the current default behavior)

mohittyagi-k replied:
done.

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_utilities/symbolic_stokes_data.h:
```diff
@@ -139,4 +139,4 @@ static int Check(const Element& rElement, const ProcessInfo& rProcessInfo)
 
```
I think we need to confirm in here that the element size formula we use is compatible with the new element instances.

rubenzorrilla replied:
I've just checked that we use the minimum element size, so there shouldn't be any problem. @mohittyagi-k forget about this comment.

rubenzorrilla commented on applications/FluidDynamicsApplication/fluid_dynamics_application.cpp:
```diff
@@ -303,6 +311,7 @@ void KratosFluidDynamicsApplication::Register() {
     KRATOS_REGISTER_CONDITION("EmbeddedAusasNavierStokesWallCondition2D2N", mEmbeddedAusasNavierStokesWallCondition2D);
     KRATOS_REGISTER_CONDITION("EmbeddedAusasNavierStokesWallCondition3D3N", mEmbeddedAusasNavierStokesWallCondition3D);
     KRATOS_REGISTER_CONDITION("StokesWallCondition3D", mStokesWallCondition3D);
+    KRATOS_REGISTER_CONDITION("StokesWallCondition3D4N", mStokesWallCondition3D4N);
```
Didn't we decide to use the Navier-Stokes condition?

mohittyagi-k replied:
I see only two NS conditions `NavierStokesWallCondition2D2N` and `NavierStokesWallCondition3D3N`. So, how would I use the NS condition for quad? Do you want me to add `NavierStokesWallCondition3D4N` also/instead?

rubenzorrilla replied:
My point is that if we have the `NavierStokesWallCondition3D4N` we can get rid of the `StokesWallCondition`. 

Maybe we can do such clean up in a future PR. And just instantiate the Stokes one as you did. I have the feeling that this one is becoming way large. What do you think?

mohittyagi-k replied:
I agree; it is worth taking the clean up in a separate PR.  Let us wrap this one up.

rubenzorrilla reviewed (commented):
Tests for the new geometries are missing. I'd also favour the use of the `NavierStokes` condition instead of the `Stokes` one.

I've no concerns about the new geometries instantiation (thanks for the work :wink:).

Commit:
default behaviour

Commit:
cleanup

Commit:
2d4n check

Commit:
2d4n stationary check

Commit:
3d6n check

Commit:
3d6n stationary check

Commit:
3d8n check

Commit:
3d8n stationary check

Commit:
refactor checks

Commit:
refactor

rubenzorrilla reviewed (dismissed):
I think this is now ready to be merged.

Things left to be done in future PR(s) are:
- Update the Stokes solver to use this new elements
- Update the Stokes solver to use the Navier-Stokes condition. Clean up the Stokes one.
- Expose the Stokes elements to the GiD GUI.
(please @mohittyagi-k @mrhashemi  @ddiezrod feel free to add things if you consider I forgot something).

@mohittyagi-k many thanks for the implementation and the patience. :wink:

Commit:
bugfix: in function call to compute normal

mohittyagi-k commented on applications/FluidDynamicsApplication/custom_conditions/stokes_wall_condition.cpp:
```diff
@@ -99,58 +127,53 @@ void StokesWallCondition<3,3>::GetDofList(DofsVectorType& rElementalDofList,
+    // Compute Unit Normal
     array_1d<double,3> Normal;
-    this->CalculateNormal(Normal); //this already contains the area
+    rGeom.Normal(Normal);
```
Sorry for the buzzkill, but this call is wrong.  This would corrupt even the stokes simulation with tet mesh. I came to know this when I tried to run a simple example as a final check. 

Another thing to take note of is that a test for Stokes wall condition is missing.  Fortunately, we have one for the Navier-Stokes wall condition.  Probably it's about time to take up the cleanup and enhancement of these conditions as the next PR.

rubenzorrilla replied:
Thanks for reporting. Let's do the upcoming changes ASAP then.
